Adds backend widgets content fixing #1307 (#1552)
plone/documentation: docs/backend/widgets.md

                           accesskey view/accesskey;
                           onselect view/onselect;
                           size view/size;
                           maxlength view/maxlength;
                           placeholder view/placeholder;
                           autocapitalize view/autocapitalize;" />
</html>
```

Now you can override the template factory in the `updateWidgets()` method of your form class.

```python
from zope.browserpage.viewpagetemplatefile import ViewPageTemplateFile as Z3ViewPageTemplateFile
from z3c.form.interfaces import INPUT_MODE

class AddForm(DefaultAddForm):

    def updateWidgets(self, prefix=None):
        """ """
        # Call parent to set-up initial widget data
        super().updateWidgets(self, prefix=prefix)

        # Note we need to be discreet to different form modes (view, edit, hidden)
        if self.fields["sections"].mode == INPUT_MODE:

            # Modify a widget with certain name for our purposes
            widget = self.widgets["sections"]

            # widget.template is a template factory -
            # Widget.render() will associate later this factory with the widget

>             widget.template = Z3ViewPageTemplateFile("templates/sections.pt")

Reviewer: Should this filename match the one give above, `yourwidget.pt`?

```

You can also interact with your `form` class instance from the widget template:

```html
<!-- Some hidden JSON data for our JavaScript by calling a method on our form class -->
<span style="display:none" tal:content="view/form/getBlockPlanJSON" />
```

### Set a template for your own widget type

You can set the template used by the widget with the `<z3c:widgetTemplate>` ZCML directive.

```xml
    <z3c:widgetTemplate
        mode="display"
        widget=".interfaces.INamedFileWidget"
        layer="z3c.form.interfaces.IFormLayer"
        template="file_display.pt"
        />
```

### Widget frame override

You can override widget templates as instructed for `z3c.form`.
`plone.app.z3cform` renders [a frame around each widget](https://github.com/plone/plone.app.z3cform/blob/master/plone/app/z3cform/templates/widget.pt), which usually consists of:

-   Label
-   Required marker
-   Description

You might want to customize this widget frame for your own form.
Below is an example of how to do it.

> Copy [`widget.pt`](https://github.com/plone/plone.app.z3cform/blob/master/plone/app/z3cform/templates/widget.pt) to your own package, and edit it.

Reviewer: The subsequent code example names a template as `demo-widget.pt`. Should this line mention that name?

Also what is the file path to the template in the package?

```python
import zope.interface
import zope.schema
from zope.schema.fieldproperty import FieldProperty

import z3c.form
from z3c.form.object import registerFactoryAdapter

class IMinMax(zope.interface.Interface):
    """ Helper schema for min and max fields """

    min = zope.schema.Float(required=False)

    max = zope.schema.Float(required=False)

@zope.interface.implementer(IMinMax)
class MinMax(object):
    """ Store min-max field values """
    min = FieldProperty(IMinMax["min"])
    max = FieldProperty(IMinMax["max"])

registerFactoryAdapter(IMinMax, MinMax)

field = zope.schema.Object(
    __name__="minmax",
    title=label,
    schema=IMinMax,
    required=False
)
```

Then you mark the widget in `updateWidgets()`:

```python
def updateWidgets(self):
    """
    """

    super(FilteringGroup, self).updateWidgets()

    # Add min and max CSS class rendering hints
    for widget in self.widgets.values():
        if isinstance(widget, z3c.form.browser.object.ObjectWidget):
            widget.template = Z3ViewPageTemplateFile("templates/minmax.pt")
            widget.addClass("min-max-widget")
            zope.interface.alsoProvides(widget, IFilterWidget)
```

> And then the page template which renders both 0. widget (min) and 1. widget (max) on the same line.

Reviewer: I have no idea what this is supposed to mean or say. Please clean this up to clarify the subsequent code example.

Author: Would this be better?

`And then create the page template minmax.pt in the same directory of your form module and paste the following code which renders both widgets (min and max) in a single row.`

> And then the page template which renders both 0. widget (min) and 1. widget (max) on the same line.

Reviewer: Ah, now that makes sense. Let's use this.

```suggestion
Then create the page template `minmax.pt` in the same directory as your form module.
Paste the following code in this file.
The code renders both widgets, {guilabel}`min` and {guilabel}`max`, in a single row.
```

Author: added suggestion

> ``` {code-block} html

Reviewer: The only reason to use `{codeblock}` is when you want to use Sphinx attributes, such as `emphasize-lines`.

````suggestion
```html
````

`updateWidgets()` binds widgets to the form's context.
{doc}`vocabularies` defined by their names are resolved at this point.

The Zope publisher will also mangle widget names based on what kind of input the widget accepts.
When an HTTP `post` request comes in, Zope publisher automatically converts `<select>` dropdowns to lists.

## Widget reference

You can find the default widgets in the browser package in `z3c.form`.
The [`z3c.form` documentation](https://z3cform.readthedocs.io/en/latest/widgets/index.html) lists all the default widgets and shows the HTML output of each.

You can find more widgets in the [`plone.app.z3cform`](https://github.com/plone/plone.app.z3cform/tree/master) package.

The main widgets are:

-   Checkbox Widget
-   Radio Widget
-   Text Widget
-   TextArea Widget
-   TextLines Widget
-   Password Widget
-   Select Widget
-   Ordered-Select Widget
-   File Widget
-   File Testing Widget
-   Image Widget
-   Multi Widget
-   Object Widget
-   Button Widget
-   Submit Widget
-   Date Widget
-   DateTime Widget
-   Time Widget
-   DateTime Picker

## Changing a field's widget

> You can change a the widget that you use for a field in several ways.

Reviewer: ```suggestion
You can change the widget that you use for a field in several ways.
```

                           onmouseover view/onmouseover;
                           onmousemove view/onmousemove;
                           onmouseout view/onmouseout;
                           onkeypress view/onkeypress;
                           onkeydown view/onkeydown;
                           onkeyup view/onkeyup;
                           value view/value;
                           disabled view/disabled;
                           tabindex view/tabindex;
                           onfocus view/onfocus;
                           onblur view/onblur;
                           onchange view/onchange;
                           readonly view/readonly;
                           alt view/alt;
                           accesskey view/accesskey;
                           onselect view/onselect;
                           size view/size;
                           maxlength view/maxlength;
                           placeholder view/placeholder;
                           autocapitalize view/autocapitalize;" />
</html>
```

Now you can override the template factory in the `updateWidgets()` method of your form class.

```python

> 

Reviewer: ```suggestion
```

        widget=".interfaces.INamedFileWidget"
        layer="z3c.form.interfaces.IFormLayer"
        template="file_display.pt"
        />
```

### Widget frame override

You can override widget templates as instructed for `z3c.form`.
`plone.app.z3cform` renders [a frame around each widget](https://github.com/plone/plone.app.z3cform/blob/master/plone/app/z3cform/templates/widget.pt), which usually consists of:

-   Label
-   Required marker
-   Description

You might want to customize this widget frame for your own form.
Below is an example of how to do it.

Copy [`widget.pt`](https://github.com/plone/plone.app.z3cform/blob/master/plone/app/z3cform/templates/widget.pt) to your own package, rename it as ``demo-widget.pt`` and edit it.

Then add the following code to `configure.zcml`.
Remember to fix the path of the template according to your own paths.

```xml
    <browser:page
        name="ploneform-render-widget"
        for=".demo.IDemoWidget"
        class="plone.app.z3cform.templates.RenderWidget"
        permission="zope.Public"
        template="path/to/template/demo-widget.pt"
        />

```

Then create a new marker interface in Python code.

```python

> 

Reviewer: ```suggestion
```

from zope.interface import Interface

class IDemoWidget(Interface):
    pass
```

Then apply this marker interface to your widgets in `form.update()`.

```python

> 

Reviewer: ```suggestion
```

from zope.interface import alsoProvides

class MyForm(...):

    def update(self):
        super().update()
        # applies custom widget frame to all widgets in this form instance
        for widget in form.widgets.values():
            alsoProvides(widget, IDemoWidget)
```

## Combined widgets

You can combine several widgets into one with `z3c.form.browser.multil.MultiWidget` and `z3c.form.browser.object.ObjectWidget` classes.

The following example shows how to create an input widget with minimum and maximum values.

```python

> 

Reviewer: ```suggestion
```

import zope.interface
import zope.schema
from zope.schema.fieldproperty import FieldProperty

import z3c.form
from z3c.form.object import registerFactoryAdapter

class IMinMax(zope.interface.Interface):
    """ Helper schema for min and max fields """

    min = zope.schema.Float(required=False)

    max = zope.schema.Float(required=False)

@zope.interface.implementer(IMinMax)
class MinMax(object):
    """ Store min-max field values """
    min = FieldProperty(IMinMax["min"])
    max = FieldProperty(IMinMax["max"])

registerFactoryAdapter(IMinMax, MinMax)

class IMyFormSchema(zope.interface.Interface):

    my_combined_field = zope.schema.Object(
        __name__="minmax",
        title=label,
        schema=IMinMax,
        required=False
    )

```

> Then you set the {guilabel}`my_combined_field` widget template in `updateWidgets()`:

Reviewer: I made a mistake. This is just an inline literal, not a GUI element.

```suggestion
Then you set the `my_combined_field` widget template in `updateWidgets()`:
```

You can find more widgets in the [`plone.app.z3cform`](https://github.com/plone/plone.app.z3cform/tree/master) package.

The main widgets are:

-   Checkbox Widget
-   Radio Widget
-   Text Widget
-   TextArea Widget
-   TextLines Widget
-   Password Widget
-   Select Widget
-   Ordered-Select Widget
-   File Widget
-   File Testing Widget
-   Image Widget
-   Multi Widget
-   Object Widget
-   Button Widget
-   Submit Widget
-   Date Widget
-   DateTime Widget
-   Time Widget
-   DateTime Picker

## Changing a field's widget

You can change a the widget that you use for a field in several ways.
This section describes these methods.

### `plone.autoform` `widget` directive

`plone.autoform` builds custom `z3c.form` forms based on a model (schema) of fields, and their widgets and options.
This model is defined as a `zope.schema` based schema, but extra hints can be supplied to control aspects of form display not normally specified in a Zope schema.

By default, `z3c.form` picks a widget based on the type of your field.
You can change the widget using the `widget` directive if you want users to enter or view data in a different format.

> For example, you can change the widget for the `human` ``boolean`` field to use "yes" and "no" radio buttons instead of its default checkbox:

Reviewer: "boolean" is not an inline literal.

Side note, we use only single backticks for inline literals in MyST or Markdown.

```suggestion
For example, you can change the widget for the `human` boolean field to use "yes" and "no" radio buttons instead of its default checkbox:
```

        self.fields["address"].widgetFactory = BlockDataGridFieldFactory
```

## Access and modify widgets

From time to time, you might need to access widgets to get their information or modify them.
For example, when a user fills out one field, a second field may need to display options dependent on the first field.

### Access a widget

You can access a widget by its field's name.

```{code-block} python
:emphasize-lines: 8
:linenos:

from z3c.form import form

class MyForm(form.Form):

    def update(self):
        super().update()
        widget = form.widgets["my_field_name"] # Get one widget

        for w in widget.items(): print(w) # Dump all widgets
```

To access a widget that's part of a group, you can't use the `updateWidgets()` method.
The groups and their widgets get initialized after the widgets have been updated.
Before that, the `groups` attribute is just a list of group factories.
During the update method though, the groups have been initialized and have their own widget list each.
To access widgets in a group, you have to access the group in the update method:

``` {code-block} python
:linenos:

> from z3c.form import form

Reviewer: Why are these added throughout the page? Line numbers don't add anything to these code examples. They just add clutter, and make `{code-block}` and an extra line break unnecessary. If you refer to a specific line number in the narrative text, or emphasize lines, then it makes sense to use the verbose form.

The above three lines (and elsewhere in this page) may be simplified as shown:

````
```python
````

Author: didn't notice your commits. fixed again

```

You can also interact with your `form` class instance from the widget template:

```html
<!-- Some hidden JSON data for our JavaScript by calling a method on our form class -->
<span style="display:none" tal:content="view/form/getBlockPlanJSON" />
```

### Set a template for your own widget type

You can set the template used by the widget with the `<z3c:widgetTemplate>` ZCML directive.

```xml
    <z3c:widgetTemplate
        mode="display"
        widget=".interfaces.INamedFileWidget"
        layer="z3c.form.interfaces.IFormLayer"
        template="file_display.pt"
        />
```

### Widget frame override

You can override widget templates as instructed for `z3c.form`.
`plone.app.z3cform` renders [a frame around each widget](https://github.com/plone/plone.app.z3cform/blob/master/plone/app/z3cform/templates/widget.pt), which usually consists of:

-   Label
-   Required marker
-   Description

You might want to customize this widget frame for your own form.
Below is an example of how to do it.

Copy [`widget.pt`](https://github.com/plone/plone.app.z3cform/blob/master/plone/app/z3cform/templates/widget.pt) to your own package, rename it as ``demo-widget.pt`` and edit it.

> Then add the following code to `configure.zcml`. Remember to fix the path of the template according to your own paths.

Reviewer: One sentence per line. 

```suggestion
Then add the following code to `configure.zcml`.
Remember to fix the path of the template according to your own paths.
```
